### PR TITLE
OH2-183 Create api of Exam Request

### PR DESCRIPTION
**Pull request by ArnaudFonzam**
See OH2-183.

### CODE REVIEW

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 1]
```diff
@@ -265,6 +265,18 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	 * Inserts one Laboratory exam {@link Laboratory} (All Procedures)
+	 *
+	 * @param laboratory - the laboratory without result
+	 * @return <code>true</code> if the exam has been inserted, <code>false</code> otherwise
+	 * @throws OHServiceException
+	 */
+	public boolean newLaboratory3(Laboratory laboratory) throws OHServiceException {
+		setPatientConsistency(laboratory);
+		return ioOperations.newLabFirstProcedure(laboratory);
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:278]: why to add this method instead to modify and use the existing one `newLaboratory`?
This method is for FirstProcedure, so also the comment is wrong (All Procedures) and therefore is not clear what are you trying to do

Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java:278]: ok, as per discussion under OH2-183, the method should be named "newExamRequest", but the suggested status LABORATORY_ACTIVE=2 is not correct, LAB_ACTIVE (this the real field name) is for future use.
Please consider to better structure the model.

Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/manager/LabManager.java:278]: @mwithi  have create the new method because all other method need exam result to be set. For this requiement we just need to set exam and the information of patient set. for the field LAB_ACTIVE, we have 3 case: LAB_ACTIVE=0  mince exam is delete LAB_ACTIVE = 1 mince exam is done and LAB_ACTIVE = 2 mince exam is requested. that is the logic

Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java]: the logic using LAB_ACTIVE as status field is wrong. Please read my comments under both OH2-183 and OHCS-84. Thanks

File: sql/step_91_oh_update_table_laboratory.sql [hunk 2]
```diff
@@ -0,0 +1,2 @@
+UNLOCK TABLES;
+ALTER TABLE LABORATORY ADD COLUMN LAB_STATUS VARCHAR(7) NULL DEFAULT NULL;
```
Comment (dbmalkovsky) [inline sql/step_91_oh_update_table_laboratory.sql:2]: Shouldn't there be an update to the demo data to go along with the adding of this column?

File: sql/step_91_oh_update_table_laboratory.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE LABORATORY ADD COLUMN LAB_STATUS VARCHAR(7) NOT NULL DEFAULT 'DONE';
```
Comment (mwithi) [inline sql/step_91_oh_update_table_laboratory.sql:1]: all statuses should be lowercase

File: sql/load_demo_data.sql [hunk 4]
```diff
@@ -1952,6 +1952,7 @@ CREATE TABLE `oh_laboratory` (
   `LAB_SEX` char(1) DEFAULT NULL,
   `LAB_MATERIAL` varchar(25) DEFAULT NULL,
   `LAB_PAT_INOUT` char(1) DEFAULT NULL,
+  `LAB_STATUS` varchar(7) DEFAULT 'DONE',
```
Comment (mwithi) [inline sql/load_demo_data.sql:1955]: statuses should be lowercase

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 5]
```diff
@@ -265,6 +265,28 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	/**
+	 * update one Laboratory exam {@link Laboratory} (All Procedures)
+	 *
+	 * @param laboratory - the laboratory without result
+	 * @return <code>true</code> if the exam has been update, <code>false</code> otherwise
+	 * @throws OHServiceException
+	 */
+	public boolean updateExamRequest(Laboratory laboratory) throws OHServiceException {
+		return ioOperations.updateLabFirstProcedure(laboratory);
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:289]: in terms of API, wouldn't be better to have a "new" method that forces the state to 'draft' and another method that just update the state?

If I want to update the laboratory (with results) there are already other methods.

Also the javadoc should be changed accordingly

Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/manager/LabManager.java:289]: yes if you want to update the laboratory with results we already have the method for that case.  about new method that force the state to 'draft' and another method that just update the state now if you want to update the state with another fields you will use other method while you can just use the updateExamRequest for all case.

Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java:289]: What I mean is that `newExamRequest` should call `laboratory.setStatus("draft")` before the ioOperations and the result should be blank even if passed, otherwise one could use the same method to save a laboratory first procedure with result directly in "done" and so duplicating the `newLabFirstProcedure` method.

Similarly, the method `updateExamRequest` should only change the laboratory status and the result should be unchanged, otherwise it may be used to update a laboratory first procedure and its result, duplicating the `updateLabFirstProcedure` method: in this case we could accept `code` and `status` as the only parameters.

In other words, a well-defined method should perform only the expected action, and the javadoc also should be more precise in this.

Also, a laboratoryStatus `enum` should be defined and used here.

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 6]
```diff
@@ -265,6 +266,40 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
 		}
 	}
 
+	/**
+	 * Inserts one Laboratory exam {@link Laboratory} (All Procedures)
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:270]:
```suggestion
	 * Inserts one Laboratory request {@link Laboratory} (All Procedures)
```

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 7]
```diff
@@ -265,6 +266,40 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	 */
+	public boolean newExamRequest(Laboratory laboratory) throws OHServiceException {
+		setPatientConsistency(laboratory);
+		laboratory.setStatus(LaboratoryStatus.draft.toString());
+		laboratory.setResult("");
+		return ioOperations.newLabFirstProcedure(laboratory);
+	}
+	
+	/**
+	 * update one Laboratory exam {@link Laboratory} (All Procedures)
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:284]:
```suggestion
	 * update one Laboratory request status (All Procedures)
```

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 8]
```diff
@@ -265,6 +266,40 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
 		}
 	}
 
+	/**
+	 * Inserts one Laboratory exam {@link Laboratory} (All Procedures)
+	 *
+	 * @param laboratory - the laboratory without result
+	 * @return <code>true</code> if the exam has been inserted, <code>false</code> otherwise
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:273]:
```suggestion
	 * @return <code>true</code> if the request has been inserted, <code>false</code> otherwise
```

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 9]
```diff
@@ -265,6 +266,40 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
 		}
 	}
 
+	/**
+	 * Inserts one Laboratory request {@link Laboratory} (All Procedures)
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:270]:
```suggestion
	 * Inserts one Laboratory request (All Procedures)
```

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 10]
```diff
@@ -265,6 +266,40 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	 */
+	public boolean newExamRequest(Laboratory laboratory) throws OHServiceException {
+		setPatientConsistency(laboratory);
+		laboratory.setStatus(LaboratoryStatus.draft.toString());
+		laboratory.setResult("");
+		return ioOperations.newLabFirstProcedure(laboratory);
+	}
+	
+	/**
+	 * update one Laboratory request {@link Laboratory} (All Procedures)
```
Comment (mwithi) [inline src/main/java/org/isf/lab/manager/LabManager.java:284]:
```suggestion
	 * update one Laboratory request (All Procedures)
```

File: src/main/java/org/isf/lab/model/Laboratory.java [hunk 11]
```diff
@@ -106,10 +106,21 @@ public class Laboratory extends Auditable<String> {
 
 	@Column(name="LAB_SEX")
 	private String sex;
+	
+	@Column(name="LAB_STATUS")
+	private String status;
```
Comment (mwithi) [inline src/main/java/org/isf/lab/model/Laboratory.java:111]: @dbmalkovsky for GUI should we assign a default value when creating a laboratory (directly in "done" I suppose)

File: src/main/java/org/isf/lab/model/Laboratory.java [hunk 12]
```diff
@@ -106,10 +106,21 @@ public class Laboratory extends Auditable<String> {
...
 	private volatile int hashCode = 0;
 
+	public String getStatus() {
+		return status;
+	}
+
+	public void setStatus(String status) {
+		this.status = status;
+	}
+
```
Comment (mwithi) [inline src/main/java/org/isf/lab/model/Laboratory.java:123]: please move getters and setters along with others in the class

File: src/main/java/org/isf/lab/model/LaboratoryStatus.java [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package org.isf.lab.model;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/model/LaboratoryStatus.java:1]: Missing copyright notice

File: src/main/java/org/isf/lab/model/LaboratoryStatus.java [hunk 14]
```diff
@@ -0,0 +1,5 @@
...
+
+public enum LaboratoryStatus {
+	draft, open, done
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]:
```suggestion
	DRAFT, OPEN, DONE
```

File: sql/step_04_all_following_steps.sql [hunk 15]
```diff
@@ -84,3 +84,4 @@ source step_87_alter_table_opd.sql;
 source step_88_alter_table_patient.sql;
 source step_89_alter_tables_innodb.sql;
 source step_90_login_retries.sql;
+source step_91_oh_update_table_laboratory.sql;
```
Comment (dbmalkovsky) [inline sql/step_04_all_following_steps.sql:87]:
```suggestion
source step_91_alter_laboratory_with_status.sql;
```
There is no need to add "_oh_" to the file name and it is better to include the changes made in the file name.  `update_table_laboratory` is too generic.

File: src/main/java/org/isf/lab/model/LaboratoryStatus.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: true... anyway in the table we should persist lowercases, how could be achieved this mapping?

**Review by dbmalkovsky — CHANGES_REQUESTED** [review]
Left various comments with code improvements.

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 11 above)
Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/model/Laboratory.java:111]: why not just create the default for status when creating the object in the constructors in `Laboratory.java`?

Comment (mwithi) [reply src/main/java/org/isf/lab/model/Laboratory.java:111]: I mean, for GUI the field is not (yet) managed, so the default (that should be "draft" in the constructor) for GUI has to be set to "done". So I think a PR on GUI is needed to set this field, wdyt?

File: src/main/java/org/isf/lab/model/LaboratoryStatus.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: Something like this?

<pre>
	public enum LaboratoryStatus {
		DRAFT("draft"), OPEN("open"), DONE("done");

		private String status;

		private LaboratoryStatus(String s) {
			status = s;
		}

		public String toString() {
			return this.status;
		}
	}
</pre>

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: yes!

[the remark this answers is not in the document]
Comment (mwithi) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: @ArnaudFonzam please use that pattern from @dbmalkovsky for enum with String values. Thanks!

File: sql/step_04_all_following_steps.sql (hunk 15 above)
Comment (mwithi) [reply sql/step_04_all_following_steps.sql:87]: suggested: `step_91_add_laboratory_status.sql`

File: src/main/java/org/isf/lab/model/LaboratoryStatus.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: okay

Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/model/LaboratoryStatus.java:4]: okay @dbmalkovsky

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 12 above)
Comment (ArnaudFonzam) [reply src/main/java/org/isf/lab/model/Laboratory.java:123]: okay

File: src/main/java/org/isf/lab/model/Laboratory.java (hunk 11 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/model/Laboratory.java:111]: Example here https://github.com/informatici/openhospital-gui/pull/1736

**Review by mwithi — DISMISSED** [review]
LGTM. Thanks a lot!

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 16]
```diff
@@ -266,6 +267,41 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	public boolean newExamRequest(Laboratory laboratory) throws OHServiceException {
+		setPatientConsistency(laboratory);
+		LaboratoryStatus laboratoryStatus = LaboratoryStatus.DRAFT;
+		laboratory.setStatus(laboratoryStatus.toString());
+		laboratory.setResult("");
+		return ioOperations.newLabFirstProcedure(laboratory);
+	}
+	
+	/**
+	 * update one Laboratory request {@link Laboratory} (All Procedures)
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/manager/LabManager.java:286]:
```suggestion
	 * Update one Laboratory request {@link Laboratory} (All Procedures)
```

File: src/main/java/org/isf/lab/manager/LabManager.java [hunk 17]
```diff
@@ -266,6 +267,41 @@ public boolean newLaboratory2(Laboratory laboratory, List<LaboratoryRow> labRow)
...
+	 * @return <code>true</code> if the request has been update, <code>false</code> otherwise
+	 * @throws OHServiceException
+	 */
+	public boolean updateExamRequest(int code, LaboratoryStatus status) throws OHServiceException {
+		Optional<Laboratory> laboratory = ioOperations.getLaboratory(code);
+		if (laboratory.isPresent()) {
+			Laboratory lab = laboratory.get();
+			lab.setStatus(status.toString());
+			return ioOperations.updateLabFirstProcedure(lab);
+		} else {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/lab/manager/LabManager.java:298]: No need for the `else` when the previous statement is a `return`

**Review by dbmalkovsky — APPROVED** [review]
LGTM.  I made two small suggestions that can be ignored if it holds up committing this PR.

File: src/main/java/org/isf/lab/manager/LabManager.java (hunk 17 above)
Comment (mwithi) [reply src/main/java/org/isf/lab/manager/LabManager.java:298]: Let's keep it this way for readability, thanks!

Comment (dbmalkovsky) [reply src/main/java/org/isf/lab/manager/LabManager.java:298]: FWIW, other methods have it without the `else`; just trying to get things consistent and without too much indentation 😏
